cou/utils/juju_utils.py

```diff
@@ -396,3 +397,101 @@ async def async_upgrade_charm(
+        """
+        self.model = model
+        self.model_name = self.model.info.name
+        self.timeout = timedelta(seconds=JujuWaiter.DEFAULT_TIMEOUT)
+        self.start_time = datetime.now()
+        self.log = logging.getLogger(self.__class__.__name__)
+
+    async def wait(self, timeout_seconds: Optional[int] = None) -> None:
+        """Wait model to stabilize.
+
+        :param timeout_seconds: wait model till timeout_seconds. If passed raise TimeoutException
+        :type timeout_seconds: Optional[int]
+        :return: None
+        :raises TimeoutException: if timeout_seconds passed.
+        :raises JujuMachineError: if it has a machine exception
+        :raises JujuAgentError: if it has an agent exception
+        :raises JujuUnitError: if it has unit exception
+        :raises JujuAppError: if it has application exception
+        """
+        self.start_time = datetime.now()
+        if timeout_seconds:
+            self.timeout = timedelta(seconds=timeout_seconds)
+        self.log.debug("Waiting to stabilize in %s seconds", self.timeout)
+
+        while True:
+            await self._ensure_model_connected()
+            try:
+                await self.model.wait_for_idle(
+                    idle_period=JujuWaiter.MODEL_IDLE_PERIOD,
+                    timeout=JujuWaiter.JUJU_IDLE_CHECK_PERIOD,
+                )
+                self.log.debug(
+                    "Model %s is idle for %s seconds.",
+                    self.model.info.name,
+                    JujuWaiter.MODEL_IDLE_PERIOD,
+                )
+                return
+            except (
+                JujuMachineError,
+                JujuAgentError,
+                JujuUnitError,
+                JujuAppError,
+            ) as ex:
+                raise ex
+            except JujuWaiter.TimeoutException as ex:
+                self.log.debug(
+                    "Unable to stabilize in %s seconds",
+                    datetime.now() - self.start_time,
+                )
+                raise ex
+            except Exception as ex:
+                # We do not care exceptions other than Juju(Machine|Agent|Unit|App)Error because
+                # when juju connection is dropped you can have wide range of exceptions depending
+                # on the case
+                self.log.debug("Unknown error while waiting to stabilize", exc_info=ex)
+
+            self._check_time()
+
+    async def _ensure_model_connected(self):
+        """Ensure that the model is connected."""
+        while _is_model_disconnected(self.model):
+            await _disconnect(self.model)
+            try:
+                self._check_time()
+                await self.model.connect_model(self.model_name)
+            except JujuWaiter.TimeoutException as ex:
+                raise ex
+            except Exception as ex:
+                self.log.debug(
+                    "Model has unexpected exception while connecting, retrying", exc_info=ex
```

Comment:
I'm not sure if this will work as expected, since you provide exc_info, but never used that in massage or fromatter?

Reply:
I've replaced with `exc_info=True`. [reference](https://docs.python.org/3.10/library/logging.html#logging.Logger.debug)